bluemira/equilibria/fem_fixed_boundary/utilities.py

```diff
@@ -59,30 +68,255 @@ class ScalarSubFunc(dolfin.UserExpression):
+    return contours
+
+
+def calculate_plasma_shape_params(psi_norm_func, mesh, psi_norm, plot=False):
+    """
+    Calculate the plasma parameters (r_geo, kappa, delta) for a given magnetic
+    isoflux using optimisation.
+
+    Parameters
+    ----------
+    psi_norm_func: callable
+        Function to calculate normalised psi
+    mesh: dolfin.Mesh
+        Mesh object to use to estimate extrema prior to optimisation
+    psi_norm: float
+        Normalised psi value for which to calculate the shape parameters
+    plot: bool
+        Whether or not to plot
+
+    Returns
+    -------
+    r_geo: float
+        Geometric major radius of the flux surface at psi_norm
+    kappa: float
+        Elongation of the flux surface at psi_norm
+    delta: float
+        Triangularity of the flux surface at psi_norm
+    """
+    points = mesh.coordinates()
+    psi_norm_array = [psi_norm_func(x) for x in points]
+
+    contour = get_tricontours(points[:, 0], points[:, 1], psi_norm_array, psi_norm)[0]
+    x, z = contour.T
+    ind_z_max = np.argmax(z)
+    pu = contour[ind_z_max]
+    ind_z_min = np.argmin(z)
+    pl = contour[ind_z_min]
+    ind_x_max = np.argmax(x)
+    po = contour[ind_x_max]
+    ind_x_min = np.argmin(x)
+    pi = contour[ind_x_min]
+
+    search_range = mesh.hmax()
+
+    def f_constrain_p95(x):
+        """
+        Constraint function for points on the psi_norm surface.
+        """
+        return psi_norm_func(x) - psi_norm
+
+    def find_extremum(func, x0):
+        """
+        Extremum finding using constrained optimisation
+        """
+        # TODO: Replace scipy minimize with something a little more robust
+        bounds = [(xi - search_range, xi + search_range) for xi in x0]
+        result = minimize(
+            func,
+            x0,
+            constraints=({"fun": f_constrain_p95, "type": "eq"}),
+            method="SLSQP",
+            bounds=bounds,
+            options={"disp": False, "ftol": 1e-10, "maxiter": 1000},
+        )
+        if not result.success:
+            bluemira_warn("Flux surface extremum finding failing:\n" f"{result.message}")
+
+        return result.x
+
+    pi_opt = find_extremum(lambda x: x[0], pi)
+
+    po_opt = find_extremum(lambda x: -x[0], po)
+
+    pl_opt = find_extremum(lambda x: x[1], pl)
+
+    pu_opt = find_extremum(lambda x: -x[1], pu)
+
+    if plot:
+        from dolfin import plot  # noqa
+
+        _, ax = plt.subplots()
+        plot(mesh)
+        ax.tricontour(points[:, 0], points[:, 1], psi_norm_array)
+        ax.plot(x, z, color="r")
+        ax.plot(*po, marker="o", color="r")
+        ax.plot(*pi, marker="o", color="r")
+        ax.plot(*pu, marker="o", color="r")
+        ax.plot(*pl, marker="o", color="r")
+
+        ax.plot(*po_opt, marker="o", color="b")
+        ax.plot(*pi_opt, marker="o", color="b")
+        ax.plot(*pu_opt, marker="o", color="b")
+        ax.plot(*pl_opt, marker="o", color="b")
+        ax.set_aspect("equal")
+        plt.show()
```

Comment:
This was mostly for debugging, I left it in in case you wanted to play around with it, but please feel free to remove.